enhancements/microshift/periodic-etcd-defragmentation-in-microshift.md

```diff
@@ -0,0 +1,168 @@
+
+#### Defragmentation from cron
+This approach would have the defragmentation condition(s) checked via a cron-job. When Microshift is installed, the cron-job(s) could be added to the tab and the system would handle the timing of the pure periodic and/or the threshold calculation.
+
+There are a few issues with this approach:
+1. It relies on cron being installed on the system and running.
+2. If Microshift is not running, the cronjob might still fire and would fail because it would be able to reach etcd which would cause a lot of false-positive failures to appear in the log.
+3. If we wanted to cover both approaches (pure periodic and threshold), then we would likely need two different cronjobs added to the tab during install.
+
+#### Defragmentation from transient systemd unit
+This would be similar to the work to launch etcd itself as a transient systemd unit. A new sub-command could be added to Microshift (or microshift-etcd) that would do much the same as the goroutine approach above: two timeouts in a loop (assuming doing both pure periodic and threshold) that will launch the defragmentation.
+Its runtime would be attached to that of Microshift and, like the goroutine, it would be easy to detect when/if a defragmentation was running or when/if it failed.
+
+This approach would be the most like an operator in OCP.
+
+### Workflow Description
+For the end user, any change we make should be completely transparent. Defragmentation is only a problem when it isn't set up (or it fails frequently) and etcd runs out of resources, causing a disruption.
+For a Microshift Administrator, they should be able to easily know when a defragmentation is occuring/occured and if it has failed.
+
+#### Variation [optional]
+TODO
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+
+TODO
+
+### Risks and Mitigations
+
+Some methods - cron-job and systemd unit - could potentially fail silently (or at least force monitoring of a non-Microshift location: cron logs/journalctl) and lead the defragmentation to not execute, which could allow etcd to grow and hit its quota which would cause further writes to fail.
+The gorountine might be superior in this case as it could send logs to the same stream as normal Microshift, which would be more likely noticed by an Administrator.
+To mitigate the above issue, we could have the systemd unit's logs streamed into the Microshift log stream; it would essentially be a subprocess of Microshift.
+
+### Drawbacks
+
+Regardless of which method is chosen, during the duration of the defragmentation, etcd will be unavailable for writes to the database; clients may still read data, but writes will be denied until the defragmentation is finished. Given the size of Microshift, this hold should not be very long, likely at most single seconds; this would be a good thing to test to ensure the disruption duration is acceptable.
```

Comment:
What behaviors have an impact on the length of time defragmentation takes? For example, would a database with many objects in it that was relatively static behave the same way as a database with few objects that were constantly being updated?

Reply:
One common theme that causes fragmentation in the context of openshift are auth tokens or build images, they are somewhat short-lived and have bigger sizes of values and can cause larger "holes" when they are deleted. 

In general, there's no easy 1:1 mapping between APIserver resources and how they translate to pages in bbolt. Rule of thumb would be to better not to create+delete objects a lot.

Reply:
Fragmentation should only occur when a write to etcd occurs; a cluster that isn't writing to etcd as much, and mainly reading, wouldn't need to defragment as much. And even then, it's possible that fragmentation doesn't occur on any individual write, because there might be enough space to store that object in a given chunk; but in general, more writes=more likely to be more fragmented.

That's why we use the fragmentation ratio in OCP, if the database isn't fragmented enough to hit our threshold, we don't do anything.

Reply:
very good point on the write aspect, the btree will try to allocate more pages sooner for faster writes, which can be later be compacted again. 

you can also check the freelist, which is used to manage all those pages: https://github.com/etcd-io/bbolt/blob/master/freelist.go
and the transaction commit logic in https://github.com/etcd-io/bbolt/blob/master/tx.go#L235-L250